client/build_context.go

```diff
@@ -0,0 +1,237 @@
+// Copyright (c) 2022, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package client
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"crypto/sha256"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// normalizePath normalizes a user-supplied path to the format supported by the io.fs package.
+func normalizePath(path string) (string, error) {
+	path, err := filepath.Abs(path)
+	if err != nil {
+		return "", err
+	}
+
+	// Paths are slash-separated.
+	path = filepath.ToSlash(path)
+
+	// Special case: the root directory is named ".".
+	if path == "/" {
+		return ".", nil
+	}
+
+	// Paths must not start with a slash.
+	path = strings.TrimPrefix(path, "/")
+
+	return path, nil
+}
+
+// writeArchive writes a compressed archive containing paths read from fsys to w.
+func writeArchive(w io.Writer, fsys fs.FS, paths []string) error {
+	gw := gzip.NewWriter(w)
+	defer gw.Close()
+
+	ar := newArchiver(fsys, gw)
+	defer ar.Close()
+
+	for _, path := range paths {
+		if err := ar.WritePath(path); err != nil {
+			return fmt.Errorf("failed to write to archive: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// getBuildContextUploadLocation obtains an upload location for a build context of the specified
+// size/hash.
+func (c *Client) getBuildContextUploadLocation(ctx context.Context, size int64, hash string) (*url.URL, error) {
+	ref := &url.URL{
+		Path: "v1/build-context",
+	}
+
+	body := struct {
+		Size     int64  `json:"size"`
+		Checksum string `json:"checksum"`
+	}{
+		Size:     size,
+		Checksum: hash,
+	}
+
+	b, err := json.Marshal(body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal request: %w", err)
+	}
+
+	req, err := c.newRequest(ctx, http.MethodPost, ref, bytes.NewReader(b))
+	if err != nil {
+		return nil, fmt.Errorf("%w", err)
+	}
+	req.Header.Set("Content-Type", "application/json")
+
+	res, err := c.httpClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("%w", err)
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode/100 != 2 { // non-2xx status code
+		return nil, fmt.Errorf("%w", errorFromResponse(res))
+	}
+
+	return url.Parse(res.Header.Get("Location"))
+}
+
+// postBuildContext uploads the build context read from r to the specified location.
+func (c *Client) postBuildContext(ctx context.Context, loc *url.URL, r io.Reader, size int64) error {
+	req, err := c.newRequest(ctx, http.MethodPost, loc, r)
```

Comment:
```suggestion
	req, err := c.newRequest(ctx, http.MethodPut, loc, r)
```

Uploads to S3 presigned URLs must use `PUT` HTTP method (https://docs.aws.amazon.com/AmazonS3/latest/userguide/PresignedUrlUploadObject.html)